### DISCUSSION TITLE
Customization over resources?

### DISCUSSION THREAD

**Discussion by ccrvlh** [rancher/rke2#3721] [Q&A] [answered]
Hi all,

i'm coming from K3S, so I might be biased.
I've been having a few issues running RKE2 within a small VM in a sandbox environment with 2vCPU and 4Gb Memory. I understand this is quite small for RKE2, but it seems that the requirements are quite high comparing to what it's actually using. At idle k9s shows 10% CPU usage and 75% memory.

![Screen Shot 2022-12-23 at 17 38 35](https://user-images.githubusercontent.com/21188280/209403604-c6d6b254-85e2-4148-a5ed-14a6ab660d71.png)

What caught my attention was the high memory usage when comparing to how much the pods are actually using (~800 Mb) and what `htop` was showing (1Gb being used), and the low ratio of CPU/Memory to Requests: apart from the api server most of the CPU/Requests% are below 10%, and Memory/Requests% sits at around ~20%. When I ran the cluster, and tried to run kata runtime daemon (no other workloads running) the cluster blew up and all components entered a restart cycle. 

I wonder if those requests numbers are actually necessary and whether it could make sense to set limits. Perhaps lower requests more in line with the resource consumption and more generous limits would work? In practice it seems that RKE2 has a 4vCPU / 8Gb minimum (I think Kata was requesting something like 300m CPU or something), which seems fairly high given the actual usage. Another options could be to allow for custom configuration of resource requirements on `config.yml`

Again, I'm coming mainly from K3S so I might be biased on how much is actually required, but I experienced lighter environments even in managed K8S solutions (DOKS mainly). I do imagine most production workflows will have way bigger environments than that, but that does restrict the possibilities, specially for sandbox/testing environments.

**Comment by brandond** [comment, marked as answer]
You might consider using K3s instead; RKE2 is not targeted towards resource-constrained environments. While you can reduce the default control-plane component requests with the `control-plane-resource-requests` option (see https://github.com/rancher/rke2/issues/3256#issuecomment-1224345989 and https://github.com/rancher/rke2/issues/2885), RKE2 may not be as good of a fit for your environment as K3s.

**Reply by ccrvlh** [reply]
Thanks @brandond.

I don't know if I would call a "resource-constrained" environment for my current setup, I'm usually working on 4c/8t 32Gb Metal (but that's relative...), but not being able to make tests in a smaller VM is definitely a disadvantage. I decided to try RKE2 mainly for it's security benefits, and still seems worth it, my main question is more in the lines of whether the hard set requirements are actually necessary.

If the main use case is huge production servers would lower requirements have any downside given that the actual requirements needs are actually a lot lower than the ones set, and the servers will handle it anyways? I guess I just don't see the downside of having lower requirements comparing to the upside / flexibility it brings when sandboxing, even not being targeted towards resource-constrained environments, the impression is that is unnecessarily restricting environments.

EDIT: Thanks for the links, will definitely take a look, I didn't find them on my previous research, so my skills are probably lacking here. Maybe just being able to have a more flexible resources configuration will do the trick for sandboxing.

**Reply by brandond** [reply]
RKE2 is targeted towards production use. The defaults are set for a baseline production cluster, not for minimal testing deployments. If you're just trying to test things in VMs, where minimal resources and a quick startup/teardown time are advantageous, I would recommend that you take a serious look at K3s. RKE2 is basically K3s with a different crypto lib and a more traditional component execution model.

**Reply by ccrvlh** [reply]
> RKE2 is basically K3s with a different crypto lib and a more traditional component execution model.

That's exactly my point.
Wouldn't it be beneficial to have lower defaults (without relevant downsides), given that setting production will also go through sandboxing and testing? As the setup differs (network, and security aspects) it seems to make more sense to test in the tool that will be used in production rather than a different one.

Anyways, thanks for the links, very helpful to be able to customize things.
I do understand the position for high defaults.

**Reply by brandond** [reply]
> Wouldn't it be beneficial to have lower defaults (without relevant downsides), given that setting production will also go through sandboxing and testing?

No, we prefer to ship with production-ready base requirements.

**Reply by ccrvlh** [reply]
Sure, I totally understand.
To be honest this is not that relevant when the customization is available, I did miss that in the docs, my apologies.
Appreciate the links and quick response!
